src/Grpc.AspNetCore.Server/Internal/PercentEncodingHelpers.cs

```diff
@@ -0,0 +1,205 @@
+
+                    var unicodeCharacters = Encoding.UTF8.GetByteCount(value.AsSpan(i, unicodeCharCount));
+                    encodedLength += unicodeCharacters * 3;
+                    i += unicodeCharCount - 1;
+                }
+                else
+                {
+                    bool unres = IsUnreservedCharacter(c);
+                    encodedLength += unres ? 1 : 3;
+                }
+            }
+
+            // Return the string if no encoding is required
+            if (value.Length == encodedLength)
+            {
+                return value;
+            }
+
+            // Encode
+            return string.Create(encodedLength, value, Encode);
+
+            static void Encode(Span<char> span, string s)
+            {
+                Span<byte> unicodeBytesBuffer = stackalloc byte[MaxUnicodeCharsReallocate * MaxUtf8BytesPerUnicodeChar];
+
+                var writePosition = 0;
+                for (var i = 0; i < s.Length; i++)
+                {
+                    var current = s[i];
+                    if (current > AsciiMaxValue)
+                    {
+                        // Get additional unicode characters
+                        var unicodeCharCount = GetUnicodeCharCount(s, i);
+
+                        var numberOfBytes = Encoding.UTF8.GetBytes(s.AsSpan(i, unicodeCharCount), unicodeBytesBuffer);
+
+                        for (var count = 0; count < numberOfBytes; count++)
+                        {
+                            EscapeAsciiChar(ref span, ref writePosition, (char)unicodeBytesBuffer[count]);
+                        }
+                        i += unicodeCharCount - 1;
+                    }
+                    else if (IsUnreservedCharacter(current))
+                    {
+                        span[writePosition++] = current;
+                    }
+                    else
+                    {
+                        EscapeAsciiChar(ref span, ref writePosition, current);
+                    }
+                }
+            }
+        }
+
+        private static void EscapeAsciiChar(ref Span<char> span, ref int writePosition, char current)
+        {
+            span[writePosition++] = '%';
+            span[writePosition++] = HexChars[current >> 4];
+            span[writePosition++] = HexChars[current & 15];
+        }
+
+        private static int GetUnicodeCharCount(string value, int currentIndex)
+        {
+            var unicodeCharCount = 0;
+            var maxSize = Math.Min(value.Length - currentIndex, MaxUnicodeCharsReallocate - 1);
+            for (; unicodeCharCount < maxSize && value[currentIndex + unicodeCharCount] > AsciiMaxValue; unicodeCharCount++)
+            {
+            }
+
+            return unicodeCharCount;
+        }
+
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        private static bool IsUnreservedCharacter(char c)
+        {
+            return ((PercentEncodingUnreservedBytes[c / 8] >> (c % 8)) & 1) != 0;
```

Comment:
how about this? https://github.com/grpc/grpc-java/blob/b363f80764bcc8452ef54511b6c6d6b596f5f177/api/src/main/java/io/grpc/Status.java#L592, seems simpler

Reply:
I copied the bitfield from the C Core implementation. I can add a comment linking to the original implementation, or replace it with Java's. Up to you.

Reply:
I think the java way is more readable and the C core approach might be a premature optimization in our case IMHO. Leaving the final decision up to you though.

Reply:
FYI they aren't equivalent. Java encodes ~ and C Core doesn't.